### DISCUSSION TITLE
Repeat characters and unable to extract tables

### DISCUSSION THREAD

**Discussion by zmingxie** [jsvine/pdfplumber#662] [Ask for help with specific PDFs]
Hi all,

I've been using this pdfplumber tool to extract data from the TREB monthly reports for personal use. It's been working great for the past reports (e.g. [April 2022](https://trreb.ca/files/market-stats/market-watch/mw2204.pdf)), but I'm having issues with its [May 2022](https://trreb.ca/files/market-stats/market-watch/mw2205.pdf) report

After some diggings, it looks like pdfplumber is having problems extracting words and table data from page 3. Here is the code snippet from my troubleshooting:

```python
import pdfplumber

pdf = pdfplumber.open("mw2205.pdf")

# The issue is with page[2] and onward and debug this page for now
p2 = pdf.pages[2]

# Draw word boxes
words = p2.extract_words(x_tolerance=0, y_tolerance=0)
im = p2.to_image()
im.reset().draw_rects(words)

# Debug table
im.reset().debug_tablefinder({})
```

* May report word racts: <img width="798" alt="image" src="https://user-images.githubusercontent.com/1136583/172061592-ae1fa39c-df35-45c4-8943-235d7b6d17ea.png">

* April report word racts: <img width="855" alt="image" src="https://user-images.githubusercontent.com/1136583/172062645-a185b96b-66ea-4bbc-8706-2e1e79dbcee9.png">

* May report table: <img width="793" alt="image" src="https://user-images.githubusercontent.com/1136583/172061609-450cfeb3-b054-451b-8fc9-27744dab8036.png">
* April report table: <img width="827" alt="image" src="https://user-images.githubusercontent.com/1136583/172062662-cbc5577a-7254-4d31-a2ef-8026003db4a7.png">

I can't figure out what changed that is causing this extraction issue. Could anyone give me some hints? Thanks!

**Comment by jsvine** [comment]
Hi @zmingxie, and very interesting! I examined `pdf.page[2].chars` for the May report, it it appears that there are a bunch of invisible, very large `A`, `b`, and `c` characters, all with `char["size"] == 33.84`. That would explain the strange artifacts you're seeing when you run `im.reset().draw_rects(words)`. You can verify this a few ways, including by inspecting `words` or `.chars`, or by opening the PDF and searching for "Abc". To achieve your goals, I'd do something like this:

```python
p2_fixed = p2.filter(lambda obj: obj.get("size", 0) < 33.84)
im2_fixed = p2_fixed.to_image()
im2_fixed.reset().debug_tablefinder()
```

Does that give you the results you expect?

**Reply by zmingxie** [reply]
@jsvine Thanks for the reply. I gave your suggestion a try, but it looks like I'm still unable to filter out all the invisible characters.
The current word count is 1238 vs 1278 earlier:

<img width="1013" alt="image" src="https://user-images.githubusercontent.com/1136583/173956221-28a390af-8daa-4dcd-858b-bbf81039e16d.png">

Is there an easy way to filter all invisible characters from a page?

**Reply by jsvine** [reply]
Whoops, try `< 33` instead of `< 33.84`; it seems I overlooked some other unusually large characters of the same kind. Doing that gets me this:

![image](https://user-images.githubusercontent.com/534702/174330399-77528dd2-db96-4403-a296-99326e8a29c2.png)

> Is there an easy way to filter all invisible characters from a page?

Unfortunately not always (or at least not in a simple way), since there are different ways that characters can *look* hidden — depending on background color, what other objects are being drawn (imagine an opaque rectangle being drawn on top of a set of characters), et cetera.

**Reply by jsvine** [reply]
Note: This seems to handle the word-extraction issue, but May report still differs from the April report with respect to the table extraction, and that seems to stem from a large number of `.rects` objects in the new report. I'd recommend looking into the different properties of those rectangles, and seeing if you can devise a filter to select only those that you really want.

**Reply by zmingxie** [reply]
I just try this again with their July report(https://trreb.ca/files/market-stats/market-watch/mw2207.pdf), and I'm still experiencing the same issue :(

Out of curiosity, if I convert this page to an image first, would it help to eliminate those invisible characters?

**Reply by jsvine** [reply]
Hi @zmingxie, and sorry to hear that they keep changing the PDF structure. Converting to an image does not modify the PDF — it's just a convenience method to help visually debug what's in the page. I suggest looking at the `page.chars` values, and using the visually debugging to understand the attributes of the characters that are giving you trouble.

**Reply by zmingxie** [reply]
Alright, I will give that a try again. Thanks for all the advice and quick responses!
